promotion.md

```diff
@@ -137,11 +144,11 @@ resources for little benefit.
 
 ### Access to a `const` or `static`
 
-When accessing a `const` in a promotable context, the restrictions on single
-assignment and named locals do not apply to the body of the `const`. All other
-restrictions, notably that the result of the `const` cannot be `Drop` or mutable
-through a reference still apply. For instance, while the previous example was
```

Comment:
What are the restrictions about "mutability through a reference here"? I would expect everything listed here to be  restrictions that already come up elsewhere in this document.

Reply:
That was a less clear way of saying interior mutability. We don't promote `&X` if `X` is a `Cell` for example. I think this should go after the colon along with `Drop` in the revised version.

As an aside, I think we used to not promote accesses to `const`s if the `const` didn't pass [some checks](https://github.com/rust-lang/rust/blob/1.38.0/src/librustc_mir/transform/qualify_consts.rs#L1567), although I think these were related to broken MIR? I was hedging here by mentioning the other restrictions since I still don't understand the reasons for that check. In any case, this will be gone once my final changes to the const-checker land in #66385.

Reply:
A `const` `Cell` should anyway be impossible -- that is a `const`-level restriction, not a promotion-level one. So I am not sure why promotion would need to check anything extra here?

Reply:
`const X: Cell<i32> = Cell::new(4)` is legal.

Reply:
> this will be gone once my final changes to the const-checker land in rust-lang/rust#66385.

They have landed, but I am not sure what exactly that means for us here,

Reply:
> const X: Cell<i32> = Cell::new(4) is legal.

Oh, right I forgot. But `&X` is not legal as the result of a `const` computation.

I think the text I originally wrote said that all the usual const restrictions still apply, did that get lost?

Reply:
I did some more editing, now it says:

> When accessing a `const` in a promotable context, its value anyway gets computed
at compile-time, so we do not have to check the initializer.  However, the
restrictions described above still apply for the *result* of the promoted
computation: in particular, it must be a valid `const` (i.e., it cannot
introduce interior mutability) and it must not require dropping.

I also moved the "technically not needed" restrictions to the bottom to give them less focus.

But I still wonder what exactly gets checked now that your PR landed, and why checking for interior mutability is not needed any more (if I understood you correctly).

Reply:
Yeah I'm sorry, I went back to my old text and it wasn't there either. Not sure what I was thinking. I added an explicit remark in this PR now.

> The qualifs in the final value of the const get checked, NeedsDrop and HasMutInterior

The drop part makes sense, for interior mutability I somehow expected that would just fall out of the usual rules for const that also have to hold for promoteds... like the example I added showed, this wouldn't even be allowed in explicit consts, let alone in promoteds. But maybe that is exactly what happens?

Reply:
I see your point. `&Cell::new(4)` is not valid in a `const` context, so it follows that `&CELL_4` is not promotable. I think this case deserves special mention because the value with interior mutability is created in a const context, but the reference is taken in a non-const one, which is subtle to me. I think the example is helpful.

Reply:
> I think this case deserves special mention because the value with interior mutability is created in a const context, but the reference is taken in a non-const one, which is subtle to me.

Agreed. Is what I have enough?

Reply:
Yes.